rmm/monitor/src/rmm/granule.rs

```diff
@@ -41,209 +41,146 @@ impl Default for GranuleState {
+    let granule = find_granule(addr);
+    let prev_state = granule.get_state();
+    match state {
+        GranuleState::Delegated => {
+            if prev_state != GranuleState::Undelegated
+                && prev_state != GranuleState::Delegated
+                && prev_state != GranuleState::RTT
+            {
+                granule.zeroize();
+                mm.unmap(addr);
+            }
+            granule.set_state(state);
+        }
+        GranuleState::RTT => granule.set_state(state),
+        GranuleState::Undelegated => unsafe {
+            // TODO: after RIPAS implement, it will be changed to panic
+            if prev_state != GranuleState::Delegated {
+                warn!(
+                    "granule[{:X}]? expected:Delegated->found:{:?}",
+                    addr, prev_state
+                );
+            }
+            granule.set_state(state);
+            GRANULE.remove(&granule.addr());
+        },
+        _ => {
+            // TODO: after RIPAS implement, it will be changed to panic
+            if prev_state != GranuleState::Delegated {
+                warn!(
+                    "granule[{:X}]? expected:Delegated->found:{:?}",
+                    addr, prev_state
+                );
+            }
+            granule.set_state(state);
+            mm.map(addr, true);
+        }
     }
+}
 
+// Define a function find_granule which returns a Granule instance from the GRANULE array using the given address and expected state.
+fn find_granule(addr: usize) -> &'static mut Granule {
+    check_validate_addr(addr);
     unsafe {
-        match GRANULE.get_mut(&idx) {
-            Some(g) => {
-                if expected_state != g.get_state() {
-                    info!(
-                        "check the {:X} granule state {:?}<-{:?}",
-                        addr,
-                        g.get_state(),
-                        expected_state
-                    );
-                }
-                Some(g)
-            }
+        match GRANULE.get_mut(&addr) {
+            Some(g) => g,
             None => {
                 let new = Granule {
-                    state: expected_state,
+                    state: GranuleState::Undelegated,
                     // refcount: 0,
-                    idx: idx,
+                    addr: addr,
                 };
-                GRANULE.insert(idx, new);
-                Some(GRANULE.get_mut(&idx).unwrap())
+                GRANULE.insert(addr, new);
+                GRANULE.get_mut(&addr).expect("granule insert failed!")
             }
         }
     }
 }
 
-// Define a function granule_addr_to_idx which returns the index of the granule using the given address.
-fn granule_addr_to_idx(addr: usize) -> usize {
-    if addr < GRANULE_BASE_ADDRESS || addr > FVP_DRAM0_END {
+// Define a function check_validate_addr which check the address is valid.
+fn check_validate_addr(addr: usize) {
```

Comment:
I think `validate_addr` is enough.

```suggestion
fn validate_addr(addr: usize) {
```

Reply:
applied it